File: server/src/routes/add_recipe_api.js
var express = require("express");
var router = express.Router();
var fs = require("fs");
const {readFileData, recipesFolder} = require("./file_utils");

router.post("/", async (req, res) => {
    try {
        let url = `https://www.themealdb.com/api/json/v1/1/lookup.php?i=${req.body.recipeId}`;
        const response = await fetch(url);
        const newRecipe = await response.json();
        const recipes = await readFileData();
        recipes.push(newRecipe.meals[0]);
        fs.writeFile(
          `${recipesFolder}/recipes.json`,
          JSON.stringify(recipes, null, 2),
          (err) => {
            if (err) throw err;
          }
        );
        res.status(200).send({ message: 'Recipe added successfully' });
      } catch (err) {
        console.log(err);
      }
});

module.exports = router;

File: server/src/modules/add_recipe_api.rs
use crate::modules::file_utils::FileUtils;
use actix_web::{web, HttpResponse};
use futures::StreamExt;
use serde_json::Value;

use super::file_utils::Recipe;

const SEARCH_API_URL: &str = "https://www.themealdb.com/api/json/v1/1/lookup.php?i=";

pub async fn add_recipe_api(mut payload: web::Payload) -> HttpResponse {
    let mut return_data: Value = Value::Null;
    while let Some(item) = payload.next().await {
        return_data = serde_json::from_slice(&item.unwrap()).unwrap();
    }
    //Check if recipe already exists
    let recipe_id = return_data["recipeId"].as_str().unwrap_or("Unknown");
    let mut recipe_json = FileUtils::read_file("recipes/recipes.json").unwrap();
    if recipe_json.iter().any(|r| r.idMeal == recipe_id) {
        return HttpResponse::Ok().body("Recipe already exists");
    }
    let recipe_data = reqwest::get(&format!("{}{}", SEARCH_API_URL, recipe_id))
        .await
        .unwrap()
        .text()
        .await
        .unwrap();

    let recipe_value: Value = serde_json::from_str(&recipe_data).unwrap();
    let new_recipe: Recipe = serde_json::from_value(recipe_value["meals"][0].clone()).unwrap();

    recipe_json.push(new_recipe);
    FileUtils::write_file("recipes/recipes.json", recipe_json).unwrap();
    HttpResponse::Ok().body("Recipe added Successfully")
}
